Readme fixes (#48)
astro-projects/astro: README.md

</h3>
<br/>

# Overview

> The `astro` library is a collection of SQL and dataframe modules that you can use to write tasks without needing to define Airflow connections and metadata for each task. By design, `astro` modules automatically pass database contexts to your tasks, meaning that you can focus on writing Python and leave metadata definitions for load time.

Reviewer: ```suggestion
The astro library is a suite of tools for writing ETL and ELT workflows in Airflow. It lets SQL engineers focus on writing SQL, Python engineers focus on writing Python, and all engineers focus on data engineering instead of configuration. By design, `astro` modules automatically pass database contexts to your tasks, meaning that you can focus on writing code and leave metadata definitions for load time.
```

Author: Wow I absolutely love this

    ```shell script
    AIRFLOW__CORE__ENABLE_XCOM_PICKLING=True
    ```

#  Using Astro as a SQL Engineer

## Schemas

By default, running `astro` for a given database creates a schema called `tmp_astro` in the database. This default behavior assumes that you have permissions to create schemas on the fly, and that there is only one user adding/removing from this schema.

For production usage, we recommend that you work with your database administrator to create a shared schema for temporary tables. This schema
can be shared across multiple users, but should be created with security in mind (e.g. don't place high security data in a shared schema).

Once you create this schema, an Airflow admin can replace `tmp_astro` by setting `AIRFLOW__ASTRO__SQL_SCHEMA="<new-temp-schema>"`, or by setting the following in `airflow.cfg`:

```bash
[astro]
sql_schema=<new-temp-schema>
```

## Setting up SQL files

When writing out a SQL DAG using `astro`, you can think of each SQL file as its own Airflow task. For example, if you wanted
to aggregate orders, aggregate customers, and then join customers and orders, you could have the following directory of files:

> ```

Reviewer: Add `ingest-models` to this folder structure

|
-- customers_table.sql
-- orders_table.sql
-- join_customers_and_orders.sql
```

> In each of these SQL files, you can create a table by having a standard `SELECT` statement. `astro` handles creating all of the temporary tables.

Reviewer: ```suggestion
In each of these SQL files, your standard `SELECT` statement automatically creates a table that can be referenced in downstream SQL files via a data dependency. `astro` handles creating all of the temporary tables required for this process.
```

### Defining dependencies

When running SQL queries in Airflow DAGs, you need to define dependencies that break up your SQL into
multiple, reproducible steps. We offer two ways to define dependencies within an `astro` SQL file:

> The first way to define a dependency is by defining a variable via the `template_vars` frontmatter argument. We set `template_vars` so that our `join_customers_and_orders` query cannot run until both our `customers_table` and `agg_orders` queries finish. This data dependency is equivalent to a task dependency in our DAG. The only difference is that we're defining it directly in our SQL instead of using Airflow's dependency operators. 

Reviewer: ```suggestion
You can define your dependency via the `template_vars` frontmatter argument. In the following example, we set `template_vars` so that our `join_customers_and_orders` query cannot run until both our `customers_table` and `agg_orders` queries finish. This data dependency is equivalent to a task dependency in our DAG. The only difference is that we're defining it directly in our SQL instead of using Airflow's dependency operators. 
```

>     )

Reviewer: Can't edit this branch, but `orders_and_customers=ingest_models["join_orders_and_customers"],` should be replaced with `**ingest_models`.

    )
```

### Raw SQL

> Most ETL use cases can be addressed by cross-sharing task outputs, as shown above with `@aql.transform`. If you need to perform SQL operation that doesn't return a table but might take a table as an argument, you can use `@aql.run_raw_sql`. 

Reviewer: ```suggestion
Most ETL use cases can be addressed by cross-sharing task outputs, as shown above with `@aql.transform`. If you need to perform a SQL operation that doesn't return a table but might take a table as an argument, you can use `@aql.run_raw_sql`. 
```

```python
@aql.run_raw_sql
def drop_table(table_to_drop):
    return "DROP TABLE IF EXISTS {table_to_drop}"
```

## Putting it All Together 

> The following is a full example DAG of a SQL + Python workflow using `astro`.

Reviewer: ```suggestion
The following is a full example DAG of a SQL + Python workflow using `astro`. We pull data from S3, run SQL transformations to merge our pulled data with existing data, and move the result of that merge into a dataframe so that we can complete complex work on it using Python / ML.
```
